src/main/java/org/reactome/addlinks/referencecreators/EnsemblBiomartMicroarrayPopulator.java

```diff
@@ -0,0 +1,192 @@
+        GKInstance refDbInst = (GKInstance) rgpInst.getAttributeValue(ReactomeJavaConstants.referenceDatabase);
+        // If reference DB is Ensembl, add the identifier associated with RGP instance.
+        if (refDbInst.getDisplayName().toLowerCase().contains("ensembl")) {
+            if (proteinToTranscripts.get(rgpIdentifier) != null) {
+                rgpProteins.add(rgpIdentifier);
+            }
+        // If reference DB is Uniprot, get Ensembl protein identifier associated with
+        // instance's identifier attribute, which should be a Uniprot identifier.
+        } else if (refDbInst.getDisplayName().toLowerCase().contains("uniprot")) {
+            if (uniprotToProtein != null && uniprotToProtein.get(rgpIdentifier) != null) {
+                rgpProteins.addAll(uniprotToProtein.get(rgpIdentifier));
+            }
+        }
+        return rgpProteins;
+    }
+
+    /**
+     * Retrieve all identifiers from RGP instances, and build a map of identifiers to all associated instances.
+     * @param rgpInstances Collection of ReferenceGeneProduct instances
+     * @return Map of RGP identifiers to associated RGP instances
+     * @throws Exception can be caused when retrieving data from GKInstance
+     */
+    private Map<String, ArrayList<GKInstance>> mapRGPIdentifiersToRGPInstances(Collection<GKInstance> rgpInstances) throws Exception {
+        Map<String, ArrayList<GKInstance>> rgpIdentifiersToRGPs = new HashMap<>();
+        for (GKInstance rgpInst : rgpInstances) {
+            String identifier = rgpInst.getAttributeValue(ReactomeJavaConstants.identifier).toString();
+            if (rgpIdentifiersToRGPs.get(identifier) != null) {
+                rgpIdentifiersToRGPs.get(identifier).add(rgpInst);
+            } else {
+                ArrayList<GKInstance> singleRGPIdentifierArray = new ArrayList<>(Arrays.asList(rgpInst));
+                rgpIdentifiersToRGPs.put(identifier, singleRGPIdentifierArray);
+            }
+        }
+        return rgpIdentifiersToRGPs;
+    }
+
+    // Sorts all microarray values added to otherIdentifier slot and updates instance with new data.
+    private void sortOtherIdentifiersAndUpdateInstance(GKInstance rgpInst) throws Exception {
+        List<String> otherIdentifiers = rgpInst.getAttributeValuesList(ReactomeJavaConstants.otherIdentifier);
+        Collections.sort(otherIdentifiers);
+        adapter.updateInstanceAttribute(rgpInst, ReactomeJavaConstants.otherIdentifier);
+    }
+
+    // Checks that none of the provided mapping structures are null.
+    private boolean allDataStructuresPopulated(Map<String, List<String>> proteinToTranscripts, Map<String, List<String>> transcriptToProbes, Map<String, List<String>> uniprotToProtein) {
+        return proteinToTranscripts != null && transcriptToProbes != null && uniprotToProtein != null;
+    }
+
+    // Get species name from json config file.
+    private Set<String> getSpeciesNames() throws IOException, ParseException {
```

Comment:
Again: is it possible to refactor the interactions with the species file into a separate utilities class?

Reply:
Done